lightningd/options.c

```diff
@@ -440,6 +440,71 @@ static char *opt_add_bind_addr(const char *arg, struct lightningd *ld)
 	return opt_add_addr_withtype(arg, ld, ADDR_LISTEN);
 }
 
+static void update_addr_field(u8 **field, const char *arg, struct lightningd *ld)
+{
+	if (*arg == '\0') {
+		*field = tal_free(*field);
+		return;
+	}
+
+	if (!*field)
+		*field = (u8 *)tal_strdup(ld, arg);
+	else {
+		u8 *new_field = (u8 *)tal_fmt(ld, "%s,%s", *field, arg);
+		tal_free(*field);
+		*field = new_field;
+	}
+
+	return;
+}
+
+static char *opt_add_alt_addr(const char *arg, struct lightningd *ld)
+{
+	assert(arg != NULL);
+
+	update_addr_field(&ld->alt_addr, arg, ld);
+	update_addr_field(&ld->alt_bind_addr, arg, ld);
+
+	return opt_add_addr_withtype(arg, ld, ADDR_LISTEN);
+}
+
+static char *opt_add_alt_bind_addr(const char *arg, struct lightningd *ld)
+{
+	assert(arg != NULL);
+
+	update_addr_field(&ld->alt_bind_addr, arg, ld);
+
+	return opt_add_addr_withtype(arg, ld, ADDR_LISTEN);
+}
+
+static char *opt_add_alt_announce_addr(const char *arg, struct lightningd *ld)
+{
+	assert(arg != NULL);
+
+	/* 'alt-bind-addr' must be set before 'alt-announce-addr'. */
+	if (ld == NULL || ld->alt_bind_addr == NULL)
+		return tal_fmt(tmpctx,
+			       "argument must first be bound by 'alt-bind-addr'");
+
+	/* Check if the argument matches any of the bound alt-bind addresses */
+	bool match_found = false;
+	char **bind_addrs = tal_strsplit(tmpctx, (char *)ld->alt_bind_addr,
+					 ",", STR_NO_EMPTY);
+	for (size_t i = 0; bind_addrs[i] != NULL; i++)
+		if (strcmp(arg, bind_addrs[i]) == 0) {
```

Comment:
This has the problem that equivalent addresses can be different strings.

There are two things here, however: restricted addresses, to which only whitelisted peers can connect to, and the addresses we advertize.  This is not clearly documented, and needs to be.